ambry-server/src/main/java/com/github/ambry/server/StatsManager.java

```diff
@@ -468,13 +471,53 @@ List<String> examineUnreachablePartitions(List<PartitionId> unreachablePartition
    */
   private class ClusterMapChangeListenerImpl implements ClusterMapChangeListener {
 
+    // We don't really care about the replica addition and removal for stats manager
     @Override
     public void onReplicaAddedOrRemoved(List<ReplicaId> addedReplicas, List<ReplicaId> removedReplicas) {
     }
 
     @Override
     public void onDataNodeRemoved(DataNodeId removedDataNode) {
       logger.info("DataNode: {} is removed from clustermap", removedDataNode);
+      if (config.enableRemoveHostFromAccountStatsStore && accountStatsStore != null
+          && removedDataNode.getDatacenterName().equals(currentNode.getDatacenterName())) {
+        // 1. There are many other server nodes that received this message and are taking action now, we have to
+        // use a distributed lock so there will be only one server node who is deleting this data node
+        // 2. Start a new thread for this task so that we don't block the listener.
+        logger.info(
+            "Receive a data node removal message for {}_{}, schedule a task to remove account stats for this host",
+            removedDataNode.getHostname(), removedDataNode.getPort());
+        scheduler.submit(() -> this.tryRemoveHostFromAccountStatsStore(removedDataNode));
+      }
+    }
+
+    private void tryRemoveHostFromAccountStatsStore(DataNodeId removedDataNode) {
+      String resource =
+          "%s_%s_%d".format(LOCK_RESOURCE_PREFIX + removedDataNode.getHostname(), removedDataNode.getPort());
+      try {
+        DistributedLock lock = clusterParticipant.getDistributedLock(resource, "AccountStatsStoreRemoval");
+        if (lock.tryLock()) {
+          logger.info("Acquired distributed lock for host removal: {}_{}", removedDataNode.getHostname(),
```

Comment:
Just a question. With distributed lock, would all hosts still execute the common code in a serialized manner?

Reply:
All the hosts would be triggered to execute this task, but the first thing in this task is to acquire lock. I set the lock timeout to be 10 minutes, which means, whoever gets the lock, will hold it for 10 minutes. All the server hosts, as long as they receive the host removal messages in 10 minute, they would fail when acquiring the lock (except for the first server host). In this way, we can be sure that there are only one server host executing this code.